src/main/java/com/microsoft/sqlserver/jdbc/ActivityCorrelator.java

```diff
@@ -8,28 +8,40 @@
 
 package com.microsoft.sqlserver.jdbc;
 
+import java.util.HashMap;
 import java.util.UUID;
 
 /**
  * ActivityCorrelator provides the APIs to access the ActivityId in TLS
  */
 final class ActivityCorrelator {
 
-    private static ThreadLocal<ActivityId> ActivityIdTls = new ThreadLocal<ActivityId>() {
-        protected ActivityId initialValue() {
-            return new ActivityId();
+    private static HashMap<Long, ActivityId> ActivityIdTlsMap = new HashMap<Long, ActivityId>();
+    
+    static void checkAndInitActivityId() {
+        long uniqueThreadId = Thread.currentThread().getId();
+        
+        //Since the Id for each thread is unique, this assures that the below code is run only once per *thread*,
+        //the first time this constructor is called.
+        if (!ActivityIdTlsMap.containsKey(uniqueThreadId)) {
+            ActivityIdTlsMap.put(uniqueThreadId, new ActivityId());
         }
-    };
+    }
 
     // Get the current ActivityId in TLS
     static ActivityId getCurrent() {
+        checkAndInitActivityId();
```

Comment:
So there's only one call for checkAndInitActivityId - you can refactor this method:
```java
static ActivityId getCurrent() {
  long uniqueThreadId = Thread.currentThread().getId();
  ActivityId id = ActivityIdTlsMap.get(uniqueThreadId);
  if (id == null) {
    id = new ActivityId();
    ActivityIdTlsMap.put(uniqueThreadId, id);
  } 
  return id;
}
```
This would save some CPU-cycles